### ISSUE TITLE
Apply changes to k3s_registries without needing to destroy cluster

### ISSUE THREAD

**Issue by timopulkkinen** [kube-hetzner/terraform-hcloud-kube-hetzner#460]
Hello, 

thanks for the project!

I made changes to k3s_registries variable in kube.tf, but these do not get applied to the cluster. 

Expected result: 
Terraform applies changes and cluster nodes get rebooted

Result:
`No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
`

Is there some workaround to apply these settings?

**Comment by mysticaltech** [comment]
@timopulkkinen This is something that needs to be done on first start. However, what would definitely work is the following. (@s3rius FYI if you have more ideas, please drop them here too).

- Create new nodepools for both control plane and agents, with the same counts as the others. Apply.
- Those new nodepools with get created with the new settings, so with the k3s_registries.
- Drain each node in the old nodepools, for each nodepool, proceed node by node, and nodepool by nodepool, set the count to 0 (don't remove), apply.
- When all is done, you will have a cluster with the same number of nodes but with the new settings.

OR

Backup your cluster with a tool like Velero from VMWare (this repo for example) and create a new cluster with the new settings, and restore everything there.


I know it's not ideal, but for now, the above is the only thing that would work IMHO.

**Comment by s3rius** [comment]
The easiest way to update is to do it manually. But this would make your terraform state out of sync with the real infra. 

@mysticaltech, I guess we can check if this variable were updated, drain all affected nodes, update the file and reboot them.

**Comment by mysticaltech** [comment]
Yep, @s3rius. something along these lines. We could use the `lifecycle` block with `create_before_destroy` to do that. Also, we can check if it exists in the state beforehand with the `terraform.state.exists(<ressource-name>)`  function.

In which case, we need to decouple the generation and copying of the registries yaml file from the rest.

**Comment by s3rius** [comment]
@mysticaltech, thanks for advice. I'll try to do it later.

**Comment by ifeulner** [comment]
All these changes lead us to *operate* the cluster... the current terraform approach is mainly for initial provisioning. Changing configurations after cluster creation needs a proper draining and deletion of the nodes, for example. If we want to support the operation of the cluster, the next question is if terraform is the right tool for it, from my perspective ? What do you thing @mysticaltech ?

**Comment by mysticaltech** [comment]
True @ifeulner, we should not operate the cluster, just deploy it. But one thing we definitely have to do better is show which variables can be changed after the initial deployment.

In that case, I think it makes sense to enable changing, because it actually touches the k3s configuration, and it's something that looks completely doable. Now it definitely poses issues for autoscaled nodes and the snapshot! 🤦

**Comment by mysticaltech** [comment]
You are right @ifeulner, maybe at some point we should provide a cli that is able to do things are is otherwise hard to pull of with terraform. In this case, we could leave things as is for now. @s3rius please, let us know what you think!

**Comment by s3rius** [comment]
Maybe it's a good approach. If I will ever create a tool to update registries, I'll write a message to this thread.

**Comment by captnCC** [comment]
@mysticaltech On the note of the autoscaler: It should be possible to include the `registries.yaml` in the `cloud-init` used for the new nodes. So we just would need to update the autoscaler manifest itself. 
For the regular nodes, a separated deployment of the the `registries.yaml` with an appropriate trigger should the be first step. And as final step using `kured` to trigger the reboot of the nodes. As per their docs placing a file `/var/run/reboot-required` should be enough to start this process 🤔

**Comment by captnCC** [comment]
@mysticaltech Working on it! But a few more problems come to my mind: 
- Existing autoscaled nodes are practically a lost case (without labels on the servers they can't be selected)
- It's all pretty unreliable/indeterministic: Since we would need to rely on `kured` to reboot the system, so there is no way to know when the changes are rolled out.

And another technical question: How does the cloud init file end up on the server after the installation of microOS?

**Comment by mysticaltech** [comment]
@captnCC Actually we touch reboot-required, kured WILL reboot the node, so that is pretty deterministic, and will do it one node at a time. Now what we could change is the `kured` sentinel period, which by default is 1h, we can reduce it to 5 minutes for instance by adding a line with   `- --period=5m` to the bottom of the kured.yaml.tpl file.

As for the cloud-config. It is read from a template file in the module and fed to data.cloudinit_config, then passed to the  user_data hetzner server param when creating the node. Basically, when microOS boots for the first time, it fetches that data.

**Comment by captnCC** [comment]
@mysticaltech Thanks for the hint about the period, that makes it much more predicable!
